cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,335 @@ namespace rapidsmpf::shuffler::detail {
+    }
 
     /**
-     * @brief Construct a new chunk of a partition.
+     * @brief Set the data buffer.
      *
-     * @param pid The ID of the partition this chunk is part of.
-     * @param cid The ID of the chunk.
-     * @param expected_num_chunks If not zero, the number of chunks of the partition
-     * expected to get from the sending rank. Ignored when it is zero.
+     * @param data The data buffer.
      */
-    Chunk(PartID pid, ChunkID cid, std::size_t expected_num_chunks);
+    void set_data_buffer(std::unique_ptr<Buffer> data) {
+        data_ = std::move(data);
+    }
 
     /**
-     * @brief Header of a metadata message.
+     * @brief Whether the data buffer is set.
+     *
+     * @return True if the data buffer is set, false otherwise.
      */
-    struct MetadataMessageHeader {
-        PartID pid;  ///< The ID of the partition this chunk is part of.
-        ChunkID cid;  ///< The ID of the chunk.
-        /// If not zero, the number of chunks of the partition expected to get from the
-        /// sending rank. Ignored when it is zero.
-        std::size_t expected_num_chunks;
-        /// If known, the size of the gpu data buffer (in bytes).
-        std::size_t gpu_data_size;
-    };
+    inline bool is_data_buffer_set() const {
+        return data_ != nullptr;
+    }
 
     /**
-     * @brief Returns a metadata message that represents this chunk.
+     * @brief Whether the metadata buffer is set.
      *
-     * @returns The metadata message as a serialized byte vector.
+     * @return True if the metadata buffer is set, false otherwise.
      */
-    [[nodiscard]] std::unique_ptr<std::vector<uint8_t>> to_metadata_message() const;
+    inline bool is_metadata_buffer_set() const {
+        return metadata_ != nullptr && !metadata_->empty();
+    }
 
     /**
-     * @brief Construct a new chunk from a metadata message.
+     * @brief Get the memory type of the data buffer.
      *
-     * @param msg A serialized metadata message previously returned by
-     * `to_metadata_message`.
-     * @returns The new chunk.
+     * @return The memory type of the data buffer.
      */
-    [[nodiscard]] static Chunk from_metadata_message(
-        std::unique_ptr<std::vector<uint8_t>> const& msg
+    inline MemoryType data_memory_type() const {
+        assert(data_);
+        return data_->mem_type();
+    }
+
+    /**
+     * @brief Get the size of the concatenated data.
+     *
+     * @return The size of the concatenated data.
+     */
+    inline size_t concat_data_size() const {
+        assert(!data_offsets_.empty());
+        return data_offsets_[n_messages() - 1];
+    }
+
+    /**
+     * @brief Get the size of the concatenated metadata.
+     *
+     * @return The size of the concatenated metadata.
+     */
+    inline size_t concat_metadata_size() const {
+        assert(metadata_);
+        assert(!meta_offsets_.empty());
+        assert(meta_offsets_[n_messages() - 1] == metadata_->size());
```

Comment:
These invariants are (or should be) enforced by the constructor, so why do we need to assert them here?

Reply:
fair point. I was mostly using the asserts to prevent me from _shooting myself in the foot_ :wink: I'll remove these